Added One-Stop RPC for the Protos Asset Store

aang114 opened the pull request:
- Currently the `owner` parameter of `get_by_tags` is commented out in Rust, as I am currently not sure how to call (and test) the Custom RPC in JavaScript (using Polkadot.js) with the `owner` parameter

Commit:
Added RPC `get_by_tags`

Commit:
Changed the type for the parameter `owner` of `get_by_tags(..,)` from `ProtoOwner` to `AccountId`

sinkingsugar commented on pallets/protos/rpc/runtime-api/src/lib.rs:
```diff
@@ -4,8 +4,21 @@ use sp_chainblocks::Hash256;
+	{
+		fn get_by_tag(tags: Tags) -> Option<Vec<Hash256>>;
+
+		fn get_by_tags(tags: Vec<Tags>, owner: Option<AccountId>, limit: u32) -> Option<Vec<Hash256>>;
```
need to add an index, so to resume iteration in the middle of the collection

sinkingsugar commented on pallets/protos/rpc/src/lib.rs:
```diff
@@ -11,9 +11,20 @@ use sp_runtime::{generic::BlockId, traits::Block as BlockT};
+	#[rpc(name = "protos_getByTags")]
+	fn get_by_tags(&self, tags: Vec<Tags>,
+				   owner: Option<AccountId>,
+				   limit: u32, from: u32, desc: bool,
```
oh here you got the right signature

sinkingsugar commented on pallets/protos/src/lib.rs:
```diff
@@ -535,6 +538,46 @@ pub mod pallet {
+				Some(owner) => {
+					if let Some(vector_protos) = <ProtosByOwner<T>>::get(ProtoOwner::User(owner)) {
+						let iter_protos = vector_protos.into_iter();
+						let iter_protos_filtered = iter_protos.filter(|proto| Self::is_proto_having_any_tags(proto, &tags));
```
does this returns another iterator? Must make sure about it or it will need to go thru ALL

aang114 replied:
Yes, this one returns an iterator

aang114 replied:
`<ProtosByOwner<T>>::get(ProtoOwner::User(owner))` does not return an iterator but `iter_protos.filter(|proto| Self::is_proto_having_any_tags(proto, &tags))` does.

Do you want `<ProtosByOwner<T>>::get(ProtoOwner::User(owner))` to return an iterator @sinkingsugar ?

sinkingsugar replied:
`get` is no good, we need to change the type like you hinted the other day into ITERABLE
https://docs.substrate.io/rustdocs/latest/frame_support/storage/trait.IterableStorageMap.html

aang114 replied:
Got it. Thanks

sinkingsugar commented on pallets/protos/src/lib.rs:
```diff
@@ -535,6 +538,46 @@ pub mod pallet {
+				},
+				None => {
+					/// TODO - Consider improving the closure in map
+					let iter_protos = tags.into_iter().map(|tag| <ProtosByTag<T>>::get(&tag).unwrap_or(Vec::new())).flatten();
```
Not efficient, you are getting all? Does this return an iterator?
I think you need to use StorageMap calls better, there are iterators maybe to investigate

aang114 replied:
Okay if we want `StorageMap` to return an iterator, I will need to change it to `IteratorStorageMap`

sinkingsugar replied:
Yes!

Commit:
1. The Maps `ProtosByTag` and `ProtosByOwner` in the pallet `protos` now uses a `StorageDoubleMap` instead of a `StorageMap` to allow for iteration
2. Implemented the Custom RPC functions `protos_getByTags` and `protos_getMetadataBatch`

sinkingsugar commented on pallets/protos/rpc/runtime-api/src/lib.rs:
```diff
@@ -4,8 +4,21 @@ use sp_chainblocks::Hash256;
+		Tags: Codec,
+		AccountId: Codec
+	{
+		fn get_by_tag(tags: Tags) -> Option<Vec<Hash256>>;
```
this is useless, we can remove

sinkingsugar commented on pallets/protos/rpc/runtime-api/src/lib.rs:
```diff
@@ -4,8 +4,21 @@ use sp_chainblocks::Hash256;
+	{
+		fn get_by_tag(tags: Tags) -> Option<Vec<Hash256>>;
+
+		fn get_by_tags(tags: Vec<Tags>, owner: Option<AccountId>, limit: u32) -> Vec<Hash256>;
```
wrong signature, where is `from` and `desc`?

sinkingsugar replied:
also `at`?

aang114 replied:
`from` and `desc` are parameters in the actual Custom RPC function (`pallets/protos/rpc/src/lib.rs`) but not in the Runtime API (because right now I'm not sure what `from` or `desc` does - could you clarify what they are supposed to? @sinkingsugar )

aang114 replied:
`at` is a parameter in in the actual Custom RPC function (`pallets/protos/rpc/src/lib.rs`)

sinkingsugar replied:
uh? we discussed...
`from` is the index to START from
and `desc` descending (vs ascending)

I see that this is not implemented, TODO

sinkingsugar commented on pallets/protos/rpc/runtime-api/src/lib.rs:
```diff
@@ -4,8 +4,21 @@ use sp_chainblocks::Hash256;
+
+		fn get_by_tags(tags: Vec<Tags>, owner: Option<AccountId>, limit: u32) -> Vec<Hash256>;
+
+		fn get_metadata_batch(batch: Vec<Hash256>, keys: Vec<Vec<u8>>) -> Vec<Option<Vec<Hash256>>>;
```
`at`?

sinkingsugar commented on pallets/protos/src/lib.rs:
```diff
@@ -134,11 +137,10 @@ pub mod pallet {
 
 	#[pallet::storage]
-	pub type ProtosByTag<T: Config> = StorageMap<_, Blake2_128Concat, Tags, Vec<Hash256>>;
+	pub type ProtosByTag<T: Config> = StorageDoubleMap<_, Blake2_128Concat, Tags, Blake2_128Concat, Hash256, bool>;
```
`Hash256` is already an hash so we can use `Identity` instead of `Blake2_128Concat` for `Hash256` types

sinkingsugar commented on pallets/protos/src/lib.rs:
```diff
@@ -134,11 +137,10 @@ pub mod pallet {
 	#[pallet::storage]
-	pub type ProtosByOwner<T: Config> =
-		StorageMap<_, Blake2_128Concat, ProtoOwner<T::AccountId>, Vec<Hash256>>;
+	pub type ProtosByOwner<T: Config> = StorageDoubleMap<_, Blake2_128Concat, ProtoOwner<T::AccountId>, Blake2_128Concat, Hash256, bool>;
```
like the above

sinkingsugar replied:
Are all those types Iterable by default? You didn't need to change type? Just a trait right?

aang114 replied:
By "types", are you referring to the Storage Items (StorageValue, StorageMap, StorageDoubelMap and StorageNMap) or to `ProtoOwner<T::AccountId>`, `Hash256` and `bool`? @sinkingsugar

sinkingsugar commented on pallets/protos/src/lib.rs:
```diff
@@ -531,8 +529,69 @@ pub mod pallet {
 
 	impl<T: Config> Pallet<T> {
+		/// This function is deprecated and will never be used
 		pub fn get_by_tag(tags: Tags) -> Option<Vec<Hash256>> {
```
remove

sinkingsugar commented on pallets/protos/src/lib.rs:
```diff
@@ -531,8 +529,69 @@ pub mod pallet {
+			Some(<ProtosByTag<T>>::iter_key_prefix(tags).collect::<Vec<Hash256>>())
+		}
+
+		fn is_proto_having_any_tags(proto_hash: &Hash256, tags: &Vec<Tags>) -> bool {
```
nit pick but better pass a slice, `&[Tags]` no?

sinkingsugar commented on pallets/protos/src/lib.rs:
```diff
@@ -531,8 +529,69 @@ pub mod pallet {
+						let vector_protos = <ProtosByTag<T>>::iter_key_prefix(tag).take(remaining as usize).collect::<Vec<Hash256>>();
+						remaining -= vector_protos.len() as u32;
+						vector_protos
+					}).flatten().take(limit as usize);
```
we apply the limit too late, if there are a million of protos we would get the full million here no?

aang114 replied:
We also apply the limit inside the `map` function (`<ProtosByTag<T>>::iter_key_prefix(tag).**take(remaining as usize)**.collect::<Vec<Hash256>>();`). 

So I don't think so? Please let me know if I'm wrong @sinkingsugar

Commit:
1. Removed deprecated Custom RPC function `get_by_tag`
2. Changed the Hashing Algorithm for the Secondary Key of the StorageDoubleMap `ProtosByTag` and `ProtosByOwner` in the pallet `protos`

sinkingsugar commented on pallets/protos/src/lib.rs:
```diff
@@ -531,8 +529,65 @@ pub mod pallet {
+				let mut vec_data_hash : Vec<Hash256> = Vec::new();
+
+				for key in &keys {
+					let data_hash = proto.metadata.get(key)?;
```
We need to handle the case where this key is not available actually, instead of `?` let's actually return a `0x0` hash? or we return `Option<Hash256>` instead of `Hash256`.
I'm asking pauline to have placeholders to render.

aang114 replied:
Oh okay. So if it returns:

1. "0x0":  It means the Proto exists but has no metadata
2. None: It means the Proto doesn't exist

aang114 replied:
I think `Option<Hash256>` is cleaner/safer

But then our return type for the function will be `Vec<Option<Vec<Option<Hash256>>>>` !

aang114 replied:
I've changed the return type in the latest commit of this branch. Have a look @sinkingsugar  (it function be further optimised in terms of dereferencing/referencing - but that's minor for now)

Commit:
Added parameters `from` and `desc` to function `get_bt_tags` (note: currently `desc` does not work)

Commit:
Changed return type of RPC `get_metadata_batch` to `Vec<Option<Vec<Option<Hash256>>>>`

Commit:
Enabled ascending order retrieval for `ProtosByTag` and `ProtosByOwner`

Commit:
Reverting back to using `StorageMap` for `ProtosByTag` and `ProtosByOwner`

sinkingsugar commented on pallets/protos/rpc/runtime-api/src/lib.rs:
```diff
@@ -10,7 +10,7 @@ sp_api::decl_runtime_apis! {
 	{
 
-		fn get_by_tags(tags: Vec<Tags>, owner: Option<AccountId>, limit: u32, from: u32, desc: bool) -> Vec<Hash256>;
+		fn get_by_tags(tags: Vec<Tags>, owner: Option<AccountId>, limit: u32, from: u32, desc: bool) -> Option<Vec<Hash256>>;
```
Empty vector not enough :) ?

aang114 replied:
I thought we only wanted "empty elements" for `get_metadata_batch()` ?

What do you mean by **"Empty vector not enough :) ?"** ?

sinkingsugar replied:
Just return an empty vector if there are not results. Why we return Option? It's pointless.

aang114 replied:
True. Changed in latest commit

sinkingsugar commented on pallets/protos/rpc/src/lib.rs:
```diff
@@ -14,7 +14,7 @@ pub use pallet_protos_rpc_runtime_api::ProtosApi as ProtosRuntimeApi;
 
 	#[rpc(name = "protos_getByTags")]
-	fn get_by_tags(&self, tags: Vec<Tags>, owner: Option<AccountId>, limit: u32, from: u32, desc: bool, at: Option<BlockHash>) ->Result<Vec<Hash256>>;
+	fn get_by_tags(&self, tags: Vec<Tags>, owner: Option<AccountId>, limit: u32, from: u32, desc: bool, at: Option<BlockHash>) ->Result<Option<Vec<Hash256>>>;
```
Empty vector

sinkingsugar commented on pallets/protos/src/lib.rs:
```diff
@@ -562,26 +553,55 @@ pub mod pallet {
+					let vector_protos = <ProtosByOwner<T>>::get(ProtoOwner::User(owner))?;
+
+					let iter_protos : Box<dyn Iterator<Item=Hash256>> = if desc {
+						Box::new(vector_protos.into_iter().rev())
```
Box is bad, why? it's a heap allocation

aang114 replied:
Fixed in latest commit. 

Please resolve this conversation if you think it has been fixed. Thanks

Commit:
Only use stack memory for `get_by_tags`

Commit:
RPC `get_by_tags`'s return type changed

Commit:
Changed `get_metadata_batch`'s function signature

Commit:
Added test cases and updated GitHub Actions

Commit:
Fixed bug

Commit:
Commented out IPFS Test from check.yml

Commit:
Changed type of `batch` in `getMetadataBatch`

Commit:
Fixed RPC test cases

Commit:
Refactored `protos` pallet's test cases

Commit:
Fixed bug

Commit:
Fixed bug

Commit:
Fixed bug

Commit:
Fixed bug

Commit:
Fixed bug

Commit:
Fixed bug

Commit:
Fixed bug

Commit:
Fixed bug

Commit:
Fixed bug

Commit:
Fixed bug

Commit:
Fixed bug

Commit:
Fixed bug

Commit:
Implemented single RPC that allows caller to get all info from Protos store

Commit:
`get_protos` RPC uses 1 parameter now instead of 7

Commit:
Fixed bug and added new test cases

sinkingsugar commented:
I will review properly on monday, can you fix conflicts with devel and resolve any solved conversation in this PR please? @aang114

Commit:
Improved readability

Commit:
Fixed bug

Commit:
Merge remote-tracking branch 'origin/devel' into rpc

# Conflicts:
#	.github/workflows/check.yml
#	pallets/protos/Cargo.toml
#	pallets/protos/rpc/Cargo.toml
#	pallets/protos/src/lib.rs
#	pallets/protos/src/tests/mod.rs
#	runtime/src/lib.rs

Commit:
Fixed bug

Commit:
Fixed bug in `check.yml`

Commit:
Fixed bug in benchmarking code

Commit:
Fixed bug

Commit:
Fixed bug

Commit:
Fixed bug in `check.yml`

Commit:
Fixed bug in `check.yml` (hopefully)

Commit:
Fixed bug in `check.yml` (hopefully)

aang114 commented:
> I will review properly on monday, can you fix conflicts with devel and resolve any solved conversation in this PR please? @aang114

Done @sinkingsugar

aang114 commented:
Btw @sinkingsugar ,  the RPCs `getByTags` and `getMetadataBatch` are now entirely redundant. So you don't need to spend any time reading them at all.

I have just left there in for the time being. Will remove them later

Commit:
Fixed bug in `check.yml`

sinkingsugar commented on .github/workflows/check.yml:
```diff
@@ -53,7 +69,90 @@ jobs:
+
+
+  #
+  # Build documentation website
```
what is all this commented stuff here?

aang114 replied:
I was trying to upload the rustdoc on NFT.Storage, but it wasn't uploading successfully. So I commented it out

Should I remove it?

sinkingsugar commented on pallets/protos/rpc/runtime-api/src/lib.rs:
```diff
@@ -3,9 +3,25 @@ use codec::Codec;
+		AccountId: Codec
+	{
+
+		fn get_by_tags(tags: Vec<Tags>, owner: Option<AccountId>, limit: u32, from: u32, desc: bool) -> Vec<Hash256>;
```
didn't we remove those 2?

aang114 replied:
> Btw @sinkingsugar , the RPCs `getByTags` and `getMetadataBatch` are now entirely redundant. So you don't need to spend any time reading them at all.
> 
> I have just left there in for the time being. Will remove them later

Not yet. But yes, I will remove them

sinkingsugar commented on pallets/protos/src/benchmarking.rs:
```diff
@@ -20,6 +20,11 @@ fn assert_last_event<T: Config>(generic_event: <T as Config>::Event) {
 
 benchmarks! {
+
+	where_clause { where
```
what does this do?

aang114 replied:
It allows us to further bound `T` (https://docs.rs/frame-benchmarking/2.0.0/frame_benchmarking/macro.benchmarks.html)

Since I bounded `T::AccountId` to `AsRef<[u8]>` in the implementation block of `Pallet<T>` (in `pallets/protos/lib.rs`), that's why we also need to bound it when benchmarking

sinkingsugar commented on pallets/protos/src/lib.rs:
```diff
@@ -64,6 +66,20 @@ pub enum ProtoOwner<TAccountId> {
+	pub desc: bool,
+	pub from: u32,
+	pub limit: u32,
+	// pub metadata_keys: Option<Vec<String>>,
```
remove?

aang114 replied:
Okay, will remove

Commit:
Removed the 2 redundant RPCs

Commit:
removed some commented out code
